File: installer/internal/webui/static/mcp-servers.js
// Seite "MCP-Server": alle Server, die die drei Assistenten in den
// Projektdateien kennen, die Pflichtliste aus K-PLAYBOOK.yaml und die
// gelesenen Dateien.
//
// Gelesen wird nur. Gemessen wird auf der Detailseite je Server, und auch
// dort erst auf Knopfdruck — diese Seite startet nichts.

const elements = {
  serversPill: document.getElementById("servers-pill"),
  serversRows: document.getElementById("servers-rows"),
  serversMessage: document.getElementById("servers-message"),
  requiredPill: document.getElementById("required-pill"),
  requiredFacts: document.getElementById("required-facts"),
  requiredGaps: document.getElementById("required-gaps"),
  requiredSnippet: document.getElementById("required-snippet"),
  requiredMessage: document.getElementById("required-message"),
  filesPill: document.getElementById("files-pill"),
  filesList: document.getElementById("files-list"),
  filesMessage: document.getElementById("files-message"),
};

// Die Spalten der Matrix, in dieser Reihenfolge. Die IDs sind die
// Pfadsegmente der Detailseite.
const ASSISTANTS = [
  { id: "claude-code", name: "Claude Code" },
  { id: "opencode", name: "OpenCode" },
  { id: "cursor", name: "Cursor" },
];

// Der Schnipsel für ein Projekt, das noch keine Pflichtliste hat.
const REQUIRED_SNIPPET = "tools:\n  mcp:\n    required:\n      - k-playbook\n";

// Muss vor den Ladefunktionen laufen: die blenden Blöcke ein, und das Menü
// zieht das nur mit, wenn es die Karten schon beobachtet.
buildBlockNav();
// Das Lebenszeichen dieses Fensters: es hält den Dienst aus dem Leerlauf und
// merkt, wenn er weg ist.
startSession((message) => {
  elements.serversMessage.textContent = message;
});
loadServers();

async function loadServers() {
  try {
    const response = await fetch("/api/mcp-servers", { cache: "no-store" });
    render(await response.json());
  } catch {
    for (const pill of [elements.serversPill, elements.requiredPill, elements.filesPill]) {
      pill.className = "pill warn";
      pill.textContent = "Nicht lesbar";
    }
    elements.serversMessage.textContent = "Die Übersicht konnte nicht geladen werden.";
  }
}

function render(data) {
  const environment = data.environment || {};
  if (!environment.installed) {
    for (const pill of [elements.serversPill, elements.requiredPill, elements.filesPill]) {
      pill.className = "pill muted";
      pill.textContent = "Nicht anwendbar";
    }
    elements.serversMessage.textContent = data.message || "";
    return;
  }

  renderMatrix(data);
  renderRequired(data);
  renderFiles(data);
}

// Die Matrix: eine Zeile je Servername, eine Spalte je Assistent. Ein Name
// kann bei OpenCode zweimal stehen, wenn beide Endungen nebeneinander liegen;
// dann zeigt die Zelle beide Einträge.
function renderMatrix(data) {
  elements.serversRows.replaceChildren();
  elements.serversMessage.textContent = "";

  const servers = data.servers || [];
  const required = new Set(data.required || []);
  const names = [...new Set([...servers.map((server) => server.name), ...required])].sort((a, b) =>
    a.localeCompare(b, "de"),
  );

  for (const name of names) {
    const row = document.createElement("tr");

    const nameCell = document.createElement("th");
    nameCell.scope = "row";
    nameCell.textContent = name;
    row.append(nameCell);

    const requiredCell = document.createElement("td");
    requiredCell.textContent = required.has(name) ? "Pflicht" : "";
    if (!required.has(name)) {
      requiredCell.className = "muted";
    }
    row.append(requiredCell);

    for (const assistant of ASSISTANTS) {
      row.append(matrixCell(name, assistant, servers, required.has(name)));
    }
    elements.serversRows.append(row);
  }

  if (names.length === 0) {
    elements.serversMessage.textContent = "In keiner der Projektdateien ist ein MCP-Server eingetragen.";
  } else if (data.message && !data.requiredError) {
    elements.serversMessage.textContent = data.message;
  }

  // Ist die Pflichtliste nicht lesbar, fehlen der Matrix die Pflichtmarken.
  // Die Pille sagt das; den Fehlertext trägt die Pflichtkarte.
  if (data.requiredError) {
    elements.serversPill.className = "pill warn";
    elements.serversPill.textContent = "Pflichtliste nicht lesbar";
    return;
  }

  const missing = data.missing || [];
  const broken = (data.files || []).filter((file) => file.error).length;
  if (missing.length > 0 || broken > 0) {
    elements.serversPill.className = "pill warn";
    elements.serversPill.textContent = missing.length > 0 ? "Pflichtserver fehlt" : "Datei nicht lesbar";
    return;
  }
  elements.serversPill.className = "pill ok";
  elements.serversPill.textContent = servers.length === 1 ? "1 Eintrag" : `${servers.length} Einträge`;
}

// Eine Zelle: der Eintrag dieses Servers bei diesem Assistenten, verlinkt auf
// die Detailseite. Fehlt er, steht ein Strich — oder „fehlt", wenn der Name
// Pflicht ist.
function matrixCell(name, assistant, servers, isRequired) {
  const cell = document.createElement("td");
  const entries = servers.filter((server) => server.name === name && server.assistantId === assistant.id);

  if (entries.length === 0) {
    if (isRequired) {
      cell.className = "warn";
      cell.textContent = "fehlt";
    } else {
      cell.className = "muted";
      cell.textContent = "—";
    }
    return cell;
  }

  for (const [index, entry] of entries.entries()) {
    if (index > 0) {
      cell.append(document.createElement("br"));
    }
    const link = document.createElement("a");
    // Die Datei kommt nur bei mehreren gleichnamigen Einträgen mit: ohne sie
    // träfe jeder Link den ersten, und der zweite wäre unerreichbar.
    link.href = detailPath(assistant.id, name, entries.length > 1 ? entry.file : "");
    link.textContent = describeEntry(entry);
    cell.append(link);
    if (!entry.enabled) {
      const note = document.createElement("span");
      note.className = "muted";
      note.textContent = " deaktiviert";
      cell.append(note);
    }
    if (entries.length > 1) {
      const file = document.createElement("span");
      file.className = "muted";
      file.textContent = ` (${entry.file})`;
      cell.append(file);
    }
  }
  if (entries.every((entry) => !entry.enabled)) {
    cell.className = "muted";
  }
  return cell;
}

// Der Eintrag in einer Zeile: Transport und das, was ihn ausmacht.
function describeEntry(entry) {
  switch (entry.transport) {
    case "local":
      return `lokal: ${[entry.command, ...(entry.args || [])].join(" ")}`;
    case "remote":
      return `remote: ${entry.url}`;
    default:
      return "unbekannte Form";
  }
}

function detailPath(assistantId, name, file) {
  const path = `/mcp-servers/${encodeURIComponent(assistantId)}/${encodeURIComponent(name)}`;
  return file ? `${path}?file=${encodeURIComponent(file)}` : path;
}

// Die Pflichtliste: was verlangt ist, wo es fehlt — oder der Schnipsel, mit
// dem ein Projekt anfängt.
function renderRequired(data) {
  elements.requiredFacts.replaceChildren();
  elements.requiredGaps.replaceChildren();
  elements.requiredMessage.textContent = "";
  elements.requiredSnippet.classList.add("hidden");

  if (data.requiredError) {
    // Ein Lesefehler an der Pflichtliste ist ihr Befund, nicht der der Matrix.
    // Erkannt am eigenen Feld: der Block steht ja da, requiredConfigured ist
    // dabei true.
    elements.requiredPill.className = "pill warn";
    elements.requiredPill.textContent = "Nicht lesbar";
    elements.requiredMessage.textContent = data.message || data.requiredError;
    return;
  }

  if (!data.requiredConfigured) {
    elements.requiredPill.className = "pill muted";
    elements.requiredPill.textContent = "Nicht festgelegt";
    elements.requiredMessage.textContent =
      "Kein Block tools.mcp.required in K-PLAYBOOK.yaml. So sähe einer aus:";
    elements.requiredSnippet.textContent = REQUIRED_SNIPPET;
    elements.requiredSnippet.classList.remove("hidden");
    return;
  }

  const required = data.required || [];
  addFact(elements.requiredFacts, "Verlangt", required.length > 0 ? required.join(", ") : "nichts (leere Liste)");

  const missing = data.missing || [];
  for (const assistant of ASSISTANTS) {
    const gaps = missing.filter((gap) => gap.assistantId === assistant.id).map((gap) => gap.name);
    const row = addFact(elements.requiredFacts, assistant.name, gaps.length > 0 ? `fehlt: ${gaps.join(", ")}` : "vollständig");
    if (gaps.length > 0) {
      row.className = "missing";
    }
  }

  elements.requiredPill.className = missing.length > 0 ? "pill warn" : "pill ok";
  elements.requiredPill.textContent = missing.length > 0 ? `${missing.length} fehlt` : "Vollständig";
}

// Die Dateien: Pfad, Assistent, Zustand, Zahl der Einträge und der Hinweis auf
// die Doppelung bei OpenCode.
function renderFiles(data) {
  elements.filesList.replaceChildren();
  elements.filesMessage.textContent = "";

  const files = data.files || [];
  let broken = 0;
  for (const file of files) {
    if (file.error) {
      broken++;
    }
    elements.filesList.append(fileBox(file));
  }

  elements.filesPill.className = broken > 0 ? "pill warn" : "pill ok";
  elements.filesPill.textContent = broken > 0 ? "Nicht lesbar" : `${files.length} Dateien`;
}

function fileBox(file) {
  const box = document.createElement("div");
  box.className = "setting";

  const head = document.createElement("div");
  head.className = "setting-head";
  const titles = document.createElement("div");
  const title = document.createElement("div");
  title.className = "setting-title";
  title.textContent = file.path;
  const state = document.createElement("div");
  state.className = file.error ? "setting-state warn" : "setting-state";
  state.textContent = file.error ? "nicht lesbar" : file.exists ? "vorhanden" : "nicht vorhanden";
  titles.append(title, state);
  head.append(titles);
  box.append(head);

  const facts = document.createElement("dl");
  facts.className = "facts";
  addFact(facts, "Assistent", file.assistant);
  addFact(facts, "Schlüssel in der Datei", file.schema);
  if (file.exists && !file.error) {
    addFact(facts, "Server", String(file.count));
  }
  if (file.error) {
    addFact(facts, "Befund", file.error);
  }
  box.append(facts);

  if (file.ambiguous) {
    const note = document.createElement("p");
    note.className = "setting-note warn";
    note.textContent =
      "opencode.json und opencode.jsonc liegen nebeneinander. OpenCode führt beide zusammen; welcher Eintrag am Ende wirkt, ist von außen nicht zu sehen. Beide werden gelesen.";
    box.append(note);
  }
  return box;
}

// Legt eine Zeile in einer Faktenliste an. Dieselbe Form wie auf den anderen
// Seiten; die Seiten teilen sich kein Skript außer session.js und nav.js.
function addFact(list, term, detail) {
  const row = document.createElement("div");
  const dt = document.createElement("dt");
  dt.textContent = term;
  const dd = document.createElement("dd");
  dd.textContent = detail;
  row.append(dt, dd);
  list.append(row);
  return row;
}

File: installer/internal/project/mcp_servers.go
package project

import (
	"fmt"
	"path/filepath"
	"sort"
)

// MCPTransport sagt, wie ein Assistent einen MCP-Server erreicht.
type MCPTransport string

const (
	// MCPTransportLocal: ein Kommando, das der Assistent selbst startet und
	// über stdin/stdout spricht.
	MCPTransportLocal MCPTransport = "local"
	// MCPTransportRemote: eine URL, hinter der der Server schon läuft — HTTP
	// oder SSE.
	MCPTransportRemote MCPTransport = "remote"
	// MCPTransportUnknown: der Eintrag passt in keine der beiden Formen. Er
	// wird gezeigt, wie er dasteht, aber nie gestartet.
	MCPTransportUnknown MCPTransport = "unknown"
)

// MCPAssistant ist einer der drei Assistenten, wie ihn die Übersicht führt:
// ID für Pfade und Vergleiche, Name für die Anzeige.
type MCPAssistant struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

// MCPAssistants sind die drei Assistenten in der Reihenfolge der Matrix.
//
// Die Namen sind dieselben, die MCPTargets in seinen Zielen trägt; die IDs
// sind die Pfadsegmente der Detailseite.
func MCPAssistants() []MCPAssistant {
	return []MCPAssistant{
		{ID: "claude-code", Name: "Claude Code"},
		{ID: "opencode", Name: "OpenCode"},
		{ID: "cursor", Name: "Cursor"},
	}
}

// MCPAssistantID übersetzt den Anzeigenamen eines Ziels in die ID. Für einen
// unbekannten Namen ist das zweite Ergebnis false.
func MCPAssistantID(name string) (string, bool) {
	for _, assistant := range MCPAssistants() {
		if assistant.Name == name {
			return assistant.ID, true
		}
	}
	return "", false
}

// MCPServerEntry ist ein Server, wie er in einer der Projektdateien steht.
//
// Die Werte aus `env` und `environment` gehören nicht in eine Antwort: dort
// stehen Tokens. Sie bleiben im unexportierten Feld und erreichen nur den
// Start der Messung; nach außen gehen ausschließlich die Schlüsselnamen.
type MCPServerEntry struct {
	Name string `json:"name"`
	// Assistant ist der Anzeigename, AssistantID das Pfadsegment.
	Assistant   string `json:"assistant"`
	AssistantID string `json:"assistantId"`
	// File ist die gelesene Datei, relativ zur Projektwurzel.
	File      string       `json:"file"`
	Schema    MCPSchema    `json:"schema"`
	Transport MCPTransport `json:"transport"`
	// Command und Args sind nur bei local gesetzt, URL nur bei remote.
	Command string   `json:"command,omitempty"`
	Args    []string `json:"args"`
	URL     string   `json:"url,omitempty"`
	// Enabled ist das OpenCode-Flag; fehlt es, gilt true. Die beiden anderen
	// Schemata kennen kein solches Flag, dort ist ein Eintrag immer aktiv.
	Enabled bool `json:"enabled"`
	// Own: der Eintrag ist die Registrierung von k-playbook selbst.
	Own bool `json:"own"`
	// Required: der Name steht in tools.mcp.required.
	Required bool `json:"required"`
	// EnvKeys sind die Schlüsselnamen aus env/environment, sortiert — ohne
	// Werte.
	EnvKeys []string `json:"envKeys"`

	env map[string]string
}

// Environ liefert env/environment als Zeilen der Form NAME=WERT, so wie ein
// Prozessstart sie braucht. Sortiert, damit zwei Starts dieselbe Umgebung
// bekommen.
func (e MCPServerEntry) Environ() []string {
	environ := make([]string, 0, len(e.env))
	for _, key := range e.EnvKeys {
		environ = append(environ, key+"="+e.env[key])
	}
	return environ
}

// MCPFileInfo ist der Zustand einer gelesenen Datei.
type MCPFileInfo struct {
	Path        string    `json:"path"`
	Assistant   string    `json:"assistant"`
	AssistantID string    `json:"assistantId"`
	Schema      MCPSchema `json:"schema"`
	Exists      bool      `json:"exists"`
	// Error ist gesetzt, wenn die Datei da, aber nicht lesbar ist. Die anderen
	// Dateien werden trotzdem gelesen.
	Error string `json:"error,omitempty"`
	// Count ist die Zahl der Server in dieser Datei.
	Count int `json:"count"`
	// Ambiguous: opencode.json und opencode.jsonc liegen nebeneinander. Beide
	// werden gelesen; welcher Eintrag am Ende wirkt, ist von außen nicht zu
	// sehen.
	Ambiguous bool `json:"ambiguous"`
}

// MCPRequirementGap ist ein Pflichtserver, der bei einem Assistenten fehlt.
type MCPRequirementGap struct {
	Name        string `json:"name"`
	Assistant   string `json:"assistant"`
	AssistantID string `json:"assistantId"`
}

// MCPServerInventory ist alles, was die Übersicht braucht: die gelesenen
// Dateien, die Server, die Pflichtliste und die Lücken darin.
type MCPServerInventory struct {
	Files   []MCPFileInfo    `json:"files"`
	Servers []MCPServerEntry `json:"servers"`
	// Required ist die Pflichtliste aus tools.mcp.required; RequiredConfigured
	// meldet, ob der Block überhaupt dastand.
	Required           []string            `json:"required"`
	RequiredConfigured bool                `json:"requiredConfigured"`
	Missing            []MCPRequirementGap `json:"missing"`
}

// mcpListTargets sind die Dateien, die für die Übersicht gelesen werden: die
// drei Registrierungsziele und, wenn beide OpenCode-Endungen nebeneinander
// liegen, auch die zweite. Geschrieben wird nur in die erste — gelesen werden
// beide, weil OpenCode beide zusammenführt.
func mcpListTargets(projectRoot string) []MCPTarget {
	targets := MCPTargets(projectRoot)
	if opencodeAmbiguous(projectRoot) {
		targets = append(targets, MCPTarget{Path: opencodeConfigJSONC, Assistant: "OpenCode", Schema: MCPSchemaOpenCode})
	}
	return targets
}

// ListMCPServers liest alle Server aus den Projektdateien der drei
// Assistenten. Eine unlesbare Datei hält die anderen nicht auf: ihr Zustand
// steht in der Dateiliste, ihre Server fehlen.
func ListMCPServers(projectRoot string) ([]MCPServerEntry, []MCPFileInfo) {
	targets := mcpListTargets(projectRoot)
	ambiguous := opencodeAmbiguous(projectRoot)

	servers := []MCPServerEntry{}
	files := make([]MCPFileInfo, 0, len(targets))
	for _, target := range targets {
		assistantID, _ := MCPAssistantID(target.Assistant)
		info := MCPFileInfo{
			Path:        target.Path,
			Assistant:   target.Assistant,
			AssistantID: assistantID,
			Schema:      target.Schema,
			Ambiguous:   target.Schema == MCPSchemaOpenCode && ambiguous,
		}

		doc, exists, err := readJSONObject(filepath.Join(projectRoot, target.Path))
		info.Exists = exists
		if err != nil {
			info.Error = err.Error()
			files = append(files, info)
			continue
		}
		section, ok := mcpSection(doc.content, target.Schema)
		if !ok {
			info.Error = string(target.Schema) + " ist kein Objekt"
			files = append(files, info)
			continue
		}

		names := make([]string, 0, len(section))
		for name := range section {
			names = append(names, name)
		}
		sort.Strings(names)
		for _, name := range names {
			servers = append(servers, describeMCPServer(target, assistantID, name, section[name]))
		}
		info.Count = len(names)
		files = append(files, info)
	}
	return servers, files
}

// describeMCPServer liest einen Eintrag tolerant: was als lokal oder remote
// erkennbar ist, wird so beschrieben; alles andere ist unknown und wird nur
// gezeigt.
//
// Claude Code und Cursor: `command` + `args` ist lokal, `type: http|sse` mit
// `url` ist remote. OpenCode: `type: local` mit `command`-Array, `type: remote`
// mit `url`; fehlt `type`, entscheidet die Form. `enabled` fehlt → true.
func describeMCPServer(target MCPTarget, assistantID string, name string, value any) MCPServerEntry {
	entry := MCPServerEntry{
		Name:        name,
		Assistant:   target.Assistant,
		AssistantID: assistantID,
		File:        target.Path,
		Schema:      target.Schema,
		Transport:   MCPTransportUnknown,
		Args:        []string{},
		Enabled:     true,
		Own:         name == MCPServerKey,
		EnvKeys:     []string{},
	}

	fields, ok := value.(map[string]any)
	if !ok {
		return entry
	}

	if enabled, ok := fields["enabled"].(bool); ok {
		entry.Enabled = enabled
	}

	kind, _ := fields["type"].(string)
	url, _ := fields["url"].(string)

	if target.Schema == MCPSchemaOpenCode {
		line, _ := fields["command"].([]any)
		switch {
		case kind == "local" || (kind == "" && len(line) > 0):
			if command, ok := firstString(line); ok {
				entry.Transport = MCPTransportLocal
				entry.Command = command
				entry.Args = stringItems(line[1:])
			}
		case kind == "remote" || (kind == "" && url != ""):
			if url != "" {
				entry.Transport = MCPTransportRemote
				entry.URL = url
			}
		}
		entry.env, entry.EnvKeys = stringMap(fields["environment"])
		return entry
	}

	command, _ := fields["command"].(string)
	switch {
	case kind == "http" || kind == "sse" || (kind == "" && command == "" && url != ""):
		if url != "" {
			entry.Transport = MCPTransportRemote
			entry.URL = url
		}
	case command != "" && (kind == "" || kind == "stdio"):
		entry.Transport = MCPTransportLocal
		entry.Command = command
		if raw, ok := fields["args"].([]any); ok {
			entry.Args = stringItems(raw)
		}
	}
	entry.env, entry.EnvKeys = stringMap(fields["env"])
	return entry
}

func firstString(values []any) (string, bool) {
	if len(values) == 0 {
		return "", false
	}
	first, ok := values[0].(string)
	return first, ok && first != ""
}

// stringItems nimmt die Zeichenketten einer Liste und übergeht den Rest: ein
// Zahlenwert in args macht den Eintrag nicht unlesbar, er fällt nur weg.
func stringItems(values []any) []string {
	items := make([]string, 0, len(values))
	for _, value := range values {
		if text, ok := value.(string); ok {
			items = append(items, text)
		}
	}
	return items
}

// stringMap liest ein Objekt aus Umgebungsvariablen: Werte als Text, Schlüssel
// sortiert. Ein anderer Typ ergibt eine leere Map.
func stringMap(value any) (map[string]string, []string) {
	fields, ok := value.(map[string]any)
	if !ok {
		return map[string]string{}, []string{}
	}
	result := make(map[string]string, len(fields))
	keys := make([]string, 0, len(fields))
	for key, raw := range fields {
		result[key] = fmt.Sprint(raw)
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return result, keys
}

// newMCPServerInventory setzt Liste und Pflichtliste zusammen: markiert die
// Pflichtserver und rechnet aus, welcher Pflichtname bei welchem Assistenten
// fehlt. Ein Eintrag in einer der beiden OpenCode-Dateien genügt.
func newMCPServerInventory(servers []MCPServerEntry, files []MCPFileInfo, required []string, configured bool) MCPServerInventory {
	inventory := MCPServerInventory{
		Files:              files,
		Servers:            servers,
		Required:           append([]string{}, required...),
		RequiredConfigured: configured,
		Missing:            []MCPRequirementGap{},
	}
	if inventory.Required == nil {
		inventory.Required = []string{}
	}

	requiredNames := map[string]bool{}
	for _, name := range required {
		requiredNames[name] = true
	}
	present := map[string]map[string]bool{}
	for index := range inventory.Servers {
		entry := &inventory.Servers[index]
		entry.Required = requiredNames[entry.Name]
		if present[entry.Name] == nil {
			present[entry.Name] = map[string]bool{}
		}
		present[entry.Name][entry.AssistantID] = true
	}

	for _, name := range required {
		for _, assistant := range MCPAssistants() {
			if !present[name][assistant.ID] {
				inventory.Missing = append(inventory.Missing, MCPRequirementGap{
					Name:        name,
					Assistant:   assistant.Name,
					AssistantID: assistant.ID,
				})
			}
		}
	}
	return inventory
}
